ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -588,6 +597,68 @@ static PrjFS_Result HandleEnumerateDirectoryRequest(const MessageHeader* request
     return callbackResult;
 }
 
+static PrjFS_Result HandleRecursivelyEnumerateDirectoryRequest(const MessageHeader* request, const char* path)
+{
+#ifdef DEBUG
+    std::cout << "PrjFSLib.HandleRecursivelyEnumerateDirectoryRequest: " << path << std::endl;
+#endif
+    
+    DIR* directory = nullptr;
+    PrjFS_Result result = PrjFS_Result_Success;
+    std::list<std::string> directoryRelativePaths;
+    directoryRelativePaths.push_back(path);
+    
+    // Walk each directory, expanding those that are found to be empty
+    char pathBuffer[PrjFSMaxPath];
+    std::list<std::string>::iterator relativePathIter = directoryRelativePaths.begin();
+    while(relativePathIter != directoryRelativePaths.end())
+    {
+        CombinePaths(s_virtualizationRootFullPath.c_str(), relativePathIter->c_str(), pathBuffer);
+    
+        // TODO(Mac): how should we handle scenarios where we were unable to fully expand the directory and
```

Comment:
Thinking about this a bit more, I'm actually not sure we have any new partial failure modes.

With this change we're still expanding folders one at a time, and if any of those expansions fails we stop attempting to expand.

I will confirm with @nickgra but I don't think this will change how the provider expands directories.  It would be no different from a user manually expanding several folders (one at a time).

The new failure mode we are introducing is the kext blocking the delete of an empty folder if the provider is not able to expand it.

Reply:
I agree, I don't think this introduces any new partial failure modes. 

If one of these expansions fails, the state in the virtualized path will be in a state where it can be repaired by retrying the I/O, even if we fail in the middle of a directory enumeration. 

From the design for atomic placeholders:
"In a case where we crash (or the machine crashes, or the user kills the mount process) while expanding a placeholder directory, we will be able to recover by calling UpdatePlaceholderIfNeeded on files that were enumerated by the previous attempt and creating the rest of the placeholders as expected."